lib/ansible/modules/cloud/vmware/vmware_guest.py

```diff
@@ -2273,6 +2282,38 @@ def wait_for_vm_ip(self, vm, poll=100, sleep=5):
 
         return facts
 
+    def get_vm_events(self, eventTypeIdList):
+        newvm = self.get_vm()
+        byEntity = vim.event.EventFilterSpec.ByEntity(entity=newvm, recursion="self")
+        filterSpec = vim.event.EventFilterSpec(entity=byEntity, eventTypeId=eventTypeIdList)
+        eventManager = self.content.eventManager
+        return eventManager.QueryEvent(filterSpec)
+
+    def wait_for_customization(self, vm, poll=10000, sleep=10):
+        facts = {}
+        thispoll = 0
+        while thispoll <= poll:
+            eventStarted = self.get_vm_events(['CustomizationStartedEvent'])
+            if len(eventStarted):
+                thispoll = 0
+                while thispoll <= poll:
+                    eventsFinishedFailed = self.get_vm_events(['CustomizationSucceeded', 'CustomizationFailed'])
+                    if len(eventsFinishedFailed):
+                        if not isinstance(eventsFinishedFailed[0], vim.event.CustomizationSucceeded):
+                            raise RuntimeError('Customization failed with error {0}:\n{1}'.format(
+                                eventsFinishedFailed[0]._wsdlName, eventsFinishedFailed[0].fullFormattedMessage))
+                        break
+                    else:
+                        time.sleep(sleep)
+                        thispoll += 1
+                newvm = self.get_vm()
+                facts = self.gather_facts(newvm)
+                return facts
```

Comment:
Why return facts, when the only place this method is called, does not use this? I'd suggest saving the effort/time to get the facts, as it will make the process longer?

Reply:
That does not take that long. The longest time is spent when waiting for the os to finish it's customization, but you're right, it is unnecessary and should thereby vanish.